### PR TITLE
Metadata map plugins

### PR DESCRIPTION
**Pull request by seth-shaw-unlv**
This PR replaces the existing strategy of using template hooks for custom metadata mappings with a plugin-based model.

On the configuration page, instead of selecting RDF, Metatag, or Custom from the drop-down; you ~~check~~ _select_ which plugin you want enabled _per metadata prefix_. 

~~Note that if multiple selected plugins implement the same metadataPrefix then the first one the code finds, wins. We might be able to make this a bit smarter by creating a check-box grid so you pick which plugins you will use for which prefix, or disabling checkboxes if the required module (e.g. RDF) is not enabled... but as a first run, it works.~~ _Update: we opted to go for select boxes per metadata prefix. This PR has been updated to reflect that change._

Now to add a metadataPrefix all you have to do is implement the plugin and select it on the config form.

Also note that switching between plugins for a given prefix will cause a LogicError the first time it is requested; but this happens with the existing template-based setup too. So it is not specific to this PR.

### CODE REVIEW

File: src/Form/RestOaiPmhSettingsForm.php [hunk 1]
```diff
@@ -145,34 +145,20 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
-      'custom' => $this->t('Custom'),
-    ];
-    $form['mapping']['mapping_source'] = [
-      '#type' => 'select',
-      '#title' => $this->t('Metadata Mapping Source'),
-      '#options' => $options,
-      '#required' => TRUE,
-      '#default_value' => $config->get('mapping_source') ? : NULL,
+    $form['mapping']['mapping_plugins'] = [
+      '#type' => 'checkboxes',
```
Comment (joecorall) [inline src/Form/RestOaiPmhSettingsForm.php]: Should this form element be of type `checkboxes`, which allows for multiple metadata mapping formats to be selected? Or of type `select` to only allow for one mapping source. I only envisioned a site admin using one mapping source, but perhaps there's a need for multiple sources?

Also, I think it should be a required form element?

**Comment by joecorall** [comment]
The <oai_dc:dc> XML element doesn't appear to have the same attributes/properties as the existing module.

Existing hook based: https://d.pr/i/AyN1tu

This new plugin approach: https://d.pr/i/Xd3SMI

I'll look into where this might have resulted from in the pull request, but I believe the <oai_dc:dc> element needs those extra attributes based on the example in the spec: http://www.openarchives.org/OAI/openarchivesprotocol.html#Record

**Comment by seth-shaw-unlv** [comment]
@joecorall, new commit should fix the oai_dc attributes.

As for the config form, I'm not wedded to a particular structure; that was simply the easiest one to do as I put this together as a proof-of-concept. 

Supposedly with this plugin structure you can have multiple modules that all provide their own plugins for various metadata prefixes. The key is finding a way to specify which plugin should be used for which metadata prefix. (E.g. I want the oai_dc from RDFMapping but I want the oai_EAD to use a plugin from another module; even though that other module provides a plugin for oai_dc I don't want to use.)

We _could_ update this form to create a list of all the prefixes available from the modules and then generate a select box for each with the available plugins for each. (This would be similar to how the entity display mode forms work which list the fields and then list the plugins available for it.) That is a bit more complicated of a form to build and validate, but we could do it.

**Comment by seth-shaw-unlv** [comment]
I almost have the metadataPrefix select boxes ready. (Meetings took me away from it.) I just have to set up the submitForm to read them and adjust the other code to use the new map in config. I hope I can get it in the morning!

**Comment by seth-shaw-unlv** [comment]
@joecorall, that should take care of the metadata formats verb response, the missing parts of the oai_dc element, and provide select boxes to choose between plugins supporting a given metadataPrefix (select 'None'to disable the metadataPrefix).

Give it a whirl and let me know what you think.

**Comment by joecorall** [comment]
Hi Seth,

This is great. Feels good for this module to be off the Drupal 7 hook approach, and onto the Drupal 8 Plugin approach. I'll add a hook_update_N() so anyone out there that might have this module installed will get updated to the slightly tweaked `rest_oai_pmh.settings` storage. Thanks so much for this contribution // pull request!

**Pull request merged**